python/ray/data/_internal/block_batching.py

```diff
@@ -361,25 +367,50 @@ def get_iter_next_batch_s_timer():
             batch = batcher.next_batch()
         yield batch
 
+    # Record batching metrics.
+    if metrics_collector is not None:
+        metrics_collector.record_metrics(batcher.get_metrics())
+
 
 def _format_batches(
     block_iter: Iterator[Block],
     batch_format: str,
     stats: Optional[Union[DatasetStats, DatasetPipelineStats]] = None,
+    metrics_collector: Optional[MetricsCollector] = None,
 ) -> Iterator[DataBatch]:
     """Given an iterator of blocks, returns an iterator of formatted batches.
 
     Args:
         block_iter: An iterator over blocks.
         batch_format: The batch format to use.
+        metrics_collector: Collector for batching metrics.
 
     Returns:
         An iterator over formatted batches.
     """
+    from ray.data._internal.arrow_block import ArrowBlockAccessor
+    from ray.data._internal.pandas_block import PandasBlockAccessor
+    from ray.data._internal.simple_block import SimpleBlockAccessor
+
+    block_type_to_zero_copy_batch_formats = {
+        ArrowBlockAccessor: {"pyarrow"},
+        PandasBlockAccessor: {"pandas", "default"},
+        SimpleBlockAccessor: {"default"},
+    }
+
+    metrics = DataMungingMetrics()
     for block in block_iter:
         with stats.iter_format_batch_s.timer() if stats else nullcontext():
-            batch = BlockAccessor.for_block(block).to_batch_format(batch_format)
+            acc = BlockAccessor.for_block(block)
```

Comment:
You may want to record the metrics from within `BlockAccessor`, e.g., `BlockAccessor(block, metrics_sink)` to avoid duplication of code recording metrics and make the code look nicer.